### ISSUE TITLE
Question: How to understand null and fdc3.nothing values during intent resolution?

### ISSUE THREAD

**Issue by ZKRobi** [finos/FDC3#1410]
### Question Area
 - [x] App Directory
 - [ ] API
 - [ ] Context Data
 - [x] Intents
 - [ ] Use Cases
 - [ ] Other

### Question
I'm tweaking our implementation around intent resolution, and some questions came up around handling fdc3.nothing and null values.
I collected my questions, and my current understanding. Please help me confirm if I understand correctly or point out if I'm mistaken.

#### In the AppDirectory interop section does "resultType": null or an omitted value mean a void IntentResult?
My understanding is yes. The only other option I thought of would be an unspecified result type that could cause unexpected errors on the client side of intents.

#### In the AppDirectory interop section is resultType: "fdc3.nothing" equivalent to "resultType": null?
In this case I have some doubts.
From the DesktopAgent/AppDirectory viewpoint yes seems like a convinient answer as if we consider fdc3.nothing and null semantically equivalent, they should both be available for resolution.
From the client point of view this leaves some uncertainity as the resolution might or might not contain a nothing context - something that should probably not cause major issues as an fdc3.nothing context holds no data.
Considering both options my understanding leans towards "yes they are equivalent" as it results in better user experience.

#### When raising an intent should "resultType": "fdc3.nothing" in the request match entries that have a null or omitted resultType field as well as ones with "resultType": "fdc3.nothing"?
Based on the above thought process my understanding is yes.

#### When raising an intent should contextType: "fdc3.nothing" match an empty or null "contexts" array as well as an array containing "fdc3.nothing" and possibly other entries?
My understanding is yes to all.

#### When raising an intent should "contextType": "fdc3.nothing" or "resultType": "fdc3.nothing" be treated as a null value for this field (and thus not filter on them)?
My understanding is no as fdc3.nothing is an explicit expression of no context.

**Comment by ZKRobi** [comment]
Thank you for the exhaustive answers and providing context.
I think I can work with this. I'll look into raising PRs for these clarifications.

**Comment by ZKRobi** [comment]
One question still came up I'd like to have confirmation on related to this:

> You can't pass null as the context arguement to raiseIntent - were you to do so I think the most appropriate response would be to reject with ResolveError.MalformedContext. Passing fdc3.nothing should match appD entries that have that type explicitly set - the Standards Working Group made the choice to not support passing nulls or undefined and instead to create fdc3.nothing to explicitly represent a lack of context. I believe this is stated, but again, think we'd be happy to review any PR raised that clarifies this decision.

If the contexts array in the AppDirectory is empty, should that be matched by providing fdc3.nothing to FindIntents?
Your answer suggests it shouldn't as it doesn't explicitly set that, but in that case there is no way to filter for that entry based on contexts.
Is this the expected behavior?

**Comment by kriswest** [comment]
@ZKRobi you raise a good point here. I (and other implementors that I'm aware of) have always assumed that the context type passed MUST be in the `interop.intents.listensFor[<intent>].contexts` array of allowed contexts for the app to be offered as a resolution option. The only exception being whereever dynamic registration is supported (adding an intent listener without an appD record or `interop.intents.listensFor` element in the record) - something that the FDC3 Workbench uses and is more useful during development. The `contexts` array is required element but the schema doesn't actually require it to contain anything (which we could fix with by adding a `minItems` clause if we wanted to).

The Standard does not currently state what should happen in this situation (nothing vs. an empty contexts array). Dynamic registration is an optional feature and also only lightly defined - we've tended to allow it where there is no `listensFor` entry for the intent that a listener was added for.

The only section of the Standard that seems relevant to this situation is:

> When raising an intent a specific context is provided as input. The type of the provided context **may** determine which applications can resolve the intent.
> 
> A context type may also be associated with multiple intents. For example, an fdc3.instrument could be associated with ViewChart, ViewNews, ViewAnalysis or other intents.
> 
> To raise an intent without a context, use the [fdc3.nothing](https://fdc3.finos.org/docs/context/ref/Nothing) context type. This type exists so that applications can explicitly declare that they support raising an intent without a context (when registering an IntentHandler or in an App Directory).

https://fdc3.finos.org/docs/api/spec#intents-and-context

Based on that I would think that it DOES make sense for passing fdc3.nothing to match to records that have a matching listensFor element for the intent, but nothing in the contexts array (although it would be better for applications to explictly declare that they support that type - as that was what was intended). If you allow that match it will not cause conformance issues for you. However, this seems poorly specified in the Standard and could be made explicit with some additional content (both in the appD schema AND API overview).
